Parse lockspace and resource names as PyBytes

aabawer opened the pull request:
WIP:
-    add missing test for all calls using lockspace/resource names
 -   change the test to use always bytes for py3
-    keep both b"name" and "name" in py2, ensuring that they are really the same.
    current vdsm code uses "name", and sanlock must be compatible with it when running on python 2.
    vdsm for python 3 will have to change all calls to use b"name".
-    If some parameters fail with py3, mark the failing parameter as xfail
 -   make the changes the module to require bytes in py3, removing possible xfails.

aabawer committed:
python: Parse lockspace and resource arguments as PyBytes

Provide converter for parsing PyBytes argument with
no-regression behavior for Py2, allowing it to accept
unicode string object within ascii chars range.

Implications:

- Py2 : no-regression in accepting both PyBytes and
	Unicodes within ascii range. TypeError will be
	raised otherwise.
- Py3 : will accept only bytes. TypeError will be raised
	otherwise.

nirs commented on tests/python_test.py:
```diff
@@ -61,22 +61,22 @@ def test_write_lockspace(tmpdir, sanlock_daemon, filename, encoding, size, offse
 
     # Test read and write with default alignment and sector size values.
-    sanlock.write_lockspace("name", path, offset=offset, iotimeout=1)
+    sanlock.write_lockspace(b"name", path, offset=offset, iotimeout=1)
```
lockspace and resource name should be parametric:
- "\xd7\x90" (py2, py3)
- u"ascii" (py2 compatibility)

We can do this for the same tests we use parametric paths.

nirs replied:
I posted this incomplete patch as example of how we should test this:
https://github.com/nirs/sanlock/commit/ff84992729c571db99b933d76226d1befec50013

@aabawer, what do you think?

nirs commented on tests/python_test.py:
```diff
@@ -147,7 +147,7 @@ def test_write_resource(tmpdir, sanlock_daemon, filename, encoding, size, offset
 
     # Test read and write with default alignment and sector size values.
-    sanlock.write_resource("ls_name", "res_name", disks)
+    sanlock.write_resource(b"ls_name", "res_name", disks)
```
Same parametric approach for resource names.

nirs commented on python/sanlock.c:
```diff
@@ -320,8 +320,9 @@ static PyObject *
-    const char *lockspace, *path;
+    const char *path;
     struct sanlk_lockspace ls;
+    PyObject *lockspace;
```
Changing the code should come after we update the tests. Doing changes before tests lead to broken tests that pass when the actual code fail.

aabawer replied:
code and test changed in same patch (squashed test here)

aabawer commented:
i have parse args tests to be added, it will address the resource name and
lockspace names as xparams

On Mon, May 20, 2019 at 12:16 PM Nir Soffer <notifications@github.com>
wrote:

> *@nirs* commented on this pull request.
> ------------------------------
>
> In tests/python_test.py
> <https://github.com/nirs/sanlock/pull/14#discussion_r285491316>:
>
> > @@ -61,22 +61,22 @@ def test_write_lockspace(tmpdir, sanlock_daemon, filename, encoding, size, offse
>      util.create_file(path, size)
>
>      # Test read and write with default alignment and sector size values.
> -    sanlock.write_lockspace("name", path, offset=offset, iotimeout=1)
> +    sanlock.write_lockspace(b"name", path, offset=offset, iotimeout=1)
>
> lockspace and resource name should be parametric:
>
>    - "\xd7\x90" (py2, py3)
>    - u"ascii" (py2 compatibility)
>
> We can do this for the same tests we use parametric paths.
> ------------------------------
>
> In tests/python_test.py
> <https://github.com/nirs/sanlock/pull/14#discussion_r285492302>:
>
> > @@ -147,7 +147,7 @@ def test_write_resource(tmpdir, sanlock_daemon, filename, encoding, size, offset
>      disks = [(path, offset)]
>
>      # Test read and write with default alignment and sector size values.
> -    sanlock.write_resource("ls_name", "res_name", disks)
> +    sanlock.write_resource(b"ls_name", "res_name", disks)
>
> Same parametric approach for resource names.
> ------------------------------
>
> In python/sanlock.c
> <https://github.com/nirs/sanlock/pull/14#discussion_r285493427>:
>
> >      struct sanlk_lockspace ls;
> +    PyObject *lockspace;
>
> Changing the code should come after we update the tests. Doing changes
> before tests lead to broken tests that pass when the actual code fail.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nirs/sanlock/pull/14?email_source=notifications&email_token=AL3BPNPH7SJ3HVUFGYLTQSLPWJT4ZA5CNFSM4HN7ALIKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOBZCIT4Q#pullrequestreview-239372786>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AL3BPNJSZEGARV2D2BOGBIDPWJT4ZANCNFSM4HN7ALIA>
> .
>

aabawer commented:
- i think you should squash both commits for bytes notation for
resoruce and lockspace. in this commit i see only the bytes notation
for lockspace

- commit msg: "Need to update mark more python 3 tests as xfai, or
include the code" --- missing "l" in xfail


On Mon, May 20, 2019 at 1:04 PM Nir Soffer <notifications@github.com> wrote:

> *@nirs* commented on this pull request.
> ------------------------------
>
> In tests/python_test.py
> <https://github.com/nirs/sanlock/pull/14#discussion_r285516238>:
>
> > @@ -61,22 +61,22 @@ def test_write_lockspace(tmpdir, sanlock_daemon, filename, encoding, size, offse
>      util.create_file(path, size)
>
>      # Test read and write with default alignment and sector size values.
> -    sanlock.write_lockspace("name", path, offset=offset, iotimeout=1)
> +    sanlock.write_lockspace(b"name", path, offset=offset, iotimeout=1)
>
> I posted this incomplete patch as example of how we should test this:
> ff84992
> <https://github.com/nirs/sanlock/commit/ff84992729c571db99b933d76226d1befec50013>
>
> @aabawer <https://github.com/aabawer>, what do you think?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nirs/sanlock/pull/14?email_source=notifications&email_token=AL3BPNOSF5KKBTEGNZGQWILPWJZTPA5CNFSM4HN7ALIKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOBZCQEBI#discussion_r285516238>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AL3BPNMXFTBWDZRWZCDQLLDPWJZTPANCNFSM4HN7ALIA>
> .
>

nirs commented:
@aabawer, this patch is incomplete example for the next version of your PR.

aabawer commented:
i have smaller tests for testing only the arg parsing for lockspace and
resoruce names.
maybe we dont want to increase the combinations of the existing tests..

On Mon, May 20, 2019 at 1:09 PM Amit Bawer <abawer@redhat.com> wrote:

> - i think you should squash both commits for bytes notation for resoruce and lockspace. in this commit i see only the bytes notation for lockspace
>
> - commit msg: "Need to update mark more python 3 tests as xfai, or include the code" --- missing "l" in xfail
>
>
> On Mon, May 20, 2019 at 1:04 PM Nir Soffer <notifications@github.com>
> wrote:
>
>> *@nirs* commented on this pull request.
>> ------------------------------
>>
>> In tests/python_test.py
>> <https://github.com/nirs/sanlock/pull/14#discussion_r285516238>:
>>
>> > @@ -61,22 +61,22 @@ def test_write_lockspace(tmpdir, sanlock_daemon, filename, encoding, size, offse
>>      util.create_file(path, size)
>>
>>      # Test read and write with default alignment and sector size values.
>> -    sanlock.write_lockspace("name", path, offset=offset, iotimeout=1)
>> +    sanlock.write_lockspace(b"name", path, offset=offset, iotimeout=1)
>>
>> I posted this incomplete patch as example of how we should test this:
>> ff84992
>> <https://github.com/nirs/sanlock/commit/ff84992729c571db99b933d76226d1befec50013>
>>
>> @aabawer <https://github.com/aabawer>, what do you think?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/nirs/sanlock/pull/14?email_source=notifications&email_token=AL3BPNOSF5KKBTEGNZGQWILPWJZTPA5CNFSM4HN7ALIKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOBZCQEBI#discussion_r285516238>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AL3BPNMXFTBWDZRWZCDQLLDPWJZTPANCNFSM4HN7ALIA>
>> .
>>
>

aabawer commented:
i think that the last two patches for changing the parsing of lockspace and resource
should be squashed together for sake of tests operating in between those as we have tests checking 
the lockspace name and resource names with same xparam and we do not want to increase combinatorial complexity.

nirs commented on tests/python_test.py:
```diff
@@ -487,3 +487,8 @@ def test_write_resource_invalid_disk(tmpdir, sanlock_daemon, disk):
+def test_killpath(tmpdir, sanlock_daemon, filename, encoding):
+    cmd_path = util.generate_path(tmpdir, filename, encoding)
+    fd = sanlock.register()
+    sanlock.killpath(cmd_path, [], fd)
```
Testing empty arguments is nice, but I think it will be more useful if we add some arguments, maybe:

    sanlock.killpath(cmd_path, [cmd_path], fd)

nirs commented on tests/python_test.py:
```diff
@@ -49,6 +51,21 @@
+    # Bytes are supported with python 2 and 3.
+    pytest.param(
+        b"\xd7\x90",
+        marks=pytest.mark.xfail(six.PY3, reason="python 3 should support only bytes")),
```
The reason is more "not implemented yet".

nirs commented on tests/python_test.py:
```diff
@@ -487,8 +504,85 @@ def test_write_resource_invalid_disk(tmpdir, sanlock_daemon, disk):
 
+
+@contextmanager
+def raises_sanlock_errno(expected_errno=errno.ECONNREFUSED):
```
This makes sense, but I remember different errors before - what did change?

aabawer replied:
Probably we had another sanlock process still running from another crashed test.
so we had our foreground test communicating with the still active process and getting the ENOENT replies.
this can be easily reproduced:
1. run tox in one terminal for all tests
2. run tox -- -k parse_args in another terminal for the relevant tests
and you'll get the different errnos by mistake.

nirs commented on tests/python_test.py:
```diff
@@ -487,8 +504,85 @@ def test_write_resource_invalid_disk(tmpdir, sanlock_daemon, disk):
+@pytest.mark.parametrize("name", LOCKSPACE_OR_RESOURCE_NAMES)
+def test_write_lockspace_parse_args(name):
+    with raises_sanlock_errno():
+        sanlock.write_lockspace(name,"ls_path")
```
missing space after ,

nirs commented on tests/python_test.py:
```diff
@@ -487,8 +504,85 @@ def test_write_resource_invalid_disk(tmpdir, sanlock_daemon, disk):
+def test_end_event_parse_args(name):
+    with raises_sanlock_errno(errno.EALREADY):
+        sanlock.end_event(-1, name)
+
```
Strange that all this fail with a different error. Maybe David can explain this error?

The empty line at the end of the file can be removed.

aabawer replied:
see prev note about sanlock running in bg from another test.

nirs commented on tests/python_test.py:
```diff
@@ -487,8 +504,85 @@ def test_write_resource_invalid_disk(tmpdir, sanlock_daemon, disk):
+@pytest.mark.parametrize("name", LOCKSPACE_OR_RESOURCE_NAMES)
+def test_rem_lockspace_parse_args(name):
+    with raises_sanlock_errno():
+        sanlock.rem_lockspace(name, 1, "ls_path", 0)
```
"ls_path" and in other tests "disk_path" are fragile, because they can match a file in the current directory.
It would be more robust to use something like "/no/path/for/you". Since we use this value about 50 times, a constant would be useful.

*

aabawer replied:
will apply dedicated patch  for dummy_path constant

nirs commented on tests/python_test.py:
```diff
@@ -54,16 +54,16 @@
-        b"\xd7\x90",
-        marks=pytest.mark.xfail(six.PY3, reason="python 3 should support only bytes")),
+        b"\xd7\x90"),
+        #marks=pytest.mark.xfail(six.PY3, reason="python 3 should support only bytes")),
```
Why comment the marks? if this works now, we can remove the pytest.param() wrapper. This is not likely to change in the future.

nirs commented on tests/python_test.py:
```diff
@@ -54,16 +54,16 @@
     pytest.param(
         u"ascii",
-        marks=pytest.mark.skipif(six.PY3, reason="python 3 supports only bytes")),
+        marks=pytest.mark.xfail(reason="python 2 does not support unicode string, python 3 only supports bytes")),
```
This is a problem, since vdsm is using unicode names with ascii content now. This change will break vdsm using this code, or any other client that happen to use unicode strings. This is pretty common, since:

```
>>> import json
>>> json.loads(json.dumps({"name": "ascii"}))
{u'name': u'ascii'}
```

I think we will need to use "O&" format, and add a function that convert unicode
object to bytes on py2, and copy bytes object otherwise. Something like:

```
/* Return new reference via addr */
static int
convert_to_pybytes(PyObject *arg, void *addr)
{
#if PY_MAJOR_VERSION == 2
    if (PyUnicode_Check(arg)) {
        PyObject *bytes = PyUnicode_AsASCIIString(arg);
        if (bytes == NULL)
            return 0;

        *(PyObject **)addr = bytes;
        return 1;
    }
#endif

    if (PyBytes_Check(arg)) {
        Py_INCREF(arg);
        *(PyObject **)addr = arg;
        return 1;
    }
    
    set type error...
    return 0; 
}
```

aabawer replied:
if we expect bytes in Py3, unicode/ascii strings would still break there; why using bytes for lockspace/resource names anyway?
This brings back to me earlier comment that we could be happy with using ascii strings for those. py3 string would support it. maybe we just need to make sure we only allow ascii strings in py3 for lockspace and resoruce name for interoperability between py2 and py3 vdsms and document this limitation in the user-guide.

aabawer replied:
we can validate post-parsing or as a convertor, utilizing the following API:
`
PyObject* PyUnicode_AsASCIIString(PyObject *unicode)
    Return value: New reference.

    Encode a Unicode object using ASCII and return the result as Python bytes object. Error handling is “strict”. Return NULL if an exception was raised by the codec.
`

aabawer replied:
we could use convert_to_pybytes shown above with same code (no ifdef) for both Py versions and not break anything on vdsm.

nirs replied:
Supporting unicode and bytes is horrible hack, needed only for compatibillity with py 2. On py 3 code must use the correct type.

If we accept unicode, we should also return it in read_xxx, and we cannot do this because py 2 allowed anything, and there is no encoding.

aabawer replied:
so this implies we'll need to straighten up the vdsm calls for lockspaces and resoruces once py3 sanlock is operational.

aabawer replied:
anyway. i think the converter should be in a separate patch so it could be easily reverted once we drop py2 for good and then we can leave the "S" formatter for py3 as planned.

aabawer committed:
tests: Apply DUMMY_PATH constant

Using hardcoded dummy paths such as "ls_path" and in other
tests "disk_path" is fragile, because those can accidently
match a file in the current directory. It is more robust
to use non-existant path.
Since we use such hard coded values about 50 times, a constant
is useful.

nirs commented on tests/util.py:
```diff
@@ -160,3 +160,18 @@ def generate_path(dirname, filename, encoding=None):
+
+def sanlock_is_running():
+    exe_file = "./src/sanlock"
+    pid_file = "/tmp/sanlock/sanlock.pid"
```
/tmp/sanlock is define in tox, we should access it here via os.environ.

nirs commented on tests/util.py:
```diff
@@ -160,3 +160,18 @@ def generate_path(dirname, filename, encoding=None):
+
+
+def sanlock_is_running():
+    exe_file = "./src/sanlock"
```
We can use SANLOCK here, or just use SANLOCK for comparing later.

nirs commented on tests/util.py:
```diff
@@ -160,3 +160,18 @@ def generate_path(dirname, filename, encoding=None):
+    pid_file = "/tmp/sanlock/sanlock.pid"
+    try:
+        with open(pid_file, "r") as f:
+            for line in f:
```
pid file contain exactly one line, so we should do:

    pid = f.readline().strip()

nirs commented on tests/util.py:
```diff
@@ -160,3 +160,18 @@ def generate_path(dirname, filename, encoding=None):
+    except OSError:
+        return False
+
+    return False
```
Would be little nicer and more correct as:

    proc_exec = "/proc/{}/exec".format(pid)

    try:
        executable = os.readlink(proc_exe)
        return os.path.samefile(executable, SANLOCK)
    except EnvironmentError as e:
        if e.errno != errno.ENOENT:
            raise

    return False

nirs commented on tests/python_test.py:
```diff
@@ -487,8 +504,94 @@ def test_write_resource_invalid_disk(tmpdir, sanlock_daemon, disk):
+
+@pytest.mark.parametrize("name", LOCKSPACE_OR_RESOURCE_NAMES)
+def test_rem_lockspace_parse_args(name):
+    assert not util.sanlock_is_running(), "Test expects sanlock to be down"
```
This will cause the test to fail, instead of error.

Why not create a no_sanlock_daemon fixture, raising SanlockIsRunning exception
if util.sanlock_is_running() returns True?

    @pytest.mark.parametrize("name", LOCKSPACE_OR_RESOURCE_NAMES)
    def test_rem_lockspace_parse_args(no_sanlock_daemon, name):
        ...

The current code duplicate the assert and the text for every test.

nirs reviewed (changes requested):
Reviewed up to:
python: Treat lockspace name argument as PyBytes

aabawer commented:
pushed PR updates.

On Tue, May 21, 2019 at 7:49 PM Nir Soffer <notifications@github.com> wrote:

> *@nirs* requested changes on this pull request.
>
> Reviewed up to:
> python: Treat lockspace name argument as PyBytes
> ------------------------------
>
> In tests/util.py
> <https://github.com/nirs/sanlock/pull/14#discussion_r286113676>:
>
> > @@ -160,3 +160,18 @@ def generate_path(dirname, filename, encoding=None):
>      if encoding:
>          path = path.encode(encoding)
>      return path
> +
> +
> +def sanlock_is_running():
> +    exe_file = "./src/sanlock"
> +    pid_file = "/tmp/sanlock/sanlock.pid"
>
> /tmp/sanlock is define in tox, we should access it here via os.environ.
> ------------------------------
>
> In tests/util.py
> <https://github.com/nirs/sanlock/pull/14#discussion_r286114228>:
>
> > @@ -160,3 +160,18 @@ def generate_path(dirname, filename, encoding=None):
>      if encoding:
>          path = path.encode(encoding)
>      return path
> +
> +
> +def sanlock_is_running():
> +    exe_file = "./src/sanlock"
>
> We can use SANLOCK here, or just use SANLOCK for comparing later.
> ------------------------------
>
> In tests/util.py
> <https://github.com/nirs/sanlock/pull/14#discussion_r286114670>:
>
> > @@ -160,3 +160,18 @@ def generate_path(dirname, filename, encoding=None):
>      if encoding:
>          path = path.encode(encoding)
>      return path
> +
> +
> +def sanlock_is_running():
> +    exe_file = "./src/sanlock"
> +    pid_file = "/tmp/sanlock/sanlock.pid"
> +    try:
> +        with open(pid_file, "r") as f:
> +            for line in f:
>
> pid file contain exactly one line, so we should do:
>
> pid = f.readline().strip()
>
> ------------------------------
>
> In tests/util.py
> <https://github.com/nirs/sanlock/pull/14#discussion_r286117924>:
>
> > @@ -160,3 +160,18 @@ def generate_path(dirname, filename, encoding=None):
>      if encoding:
>          path = path.encode(encoding)
>      return path
> +
> +
> +def sanlock_is_running():
> +    exe_file = "./src/sanlock"
> +    pid_file = "/tmp/sanlock/sanlock.pid"
> +    try:
> +        with open(pid_file, "r") as f:
> +            for line in f:
> +                pid = line.strip()
> +                if os.path.samefile(os.readlink("/proc/"+str(pid)+"/exe"), exe_file):
> +                    return True
> +    except OSError:
> +        return False
> +
> +    return False
>
> Would be little nicer and more correct as:
>
> proc_exec = "/proc/{}/exec".format(pid)
>
> try:
>     executable = os.readlink(proc_exe)
>     return os.path.samefile(executable, SANLOCK)
> except EnvironmentError as e:
>     if e.errno != errno.ENOENT:
>         raise
>
> return False
>
> ------------------------------
>
> In tests/python_test.py
> <https://github.com/nirs/sanlock/pull/14#discussion_r286119939>:
>
> >  @pytest.mark.parametrize("filename,encoding", FILE_NAMES)
>  def test_killpath(tmpdir, sanlock_daemon, filename, encoding):
>      cmd_path = util.generate_path(tmpdir, filename, encoding)
>      fd = sanlock.register()
>      sanlock.killpath(cmd_path, [cmd_path], fd)
> +
> +
> +@contextmanager
> +def raises_sanlock_errno(expected_errno=errno.ECONNREFUSED):
> +    with pytest.raises(sanlock.SanlockException) as e:
> +        yield
> +    assert e.value.errno == expected_errno
> +
> +
> +@pytest.mark.parametrize("name", LOCKSPACE_OR_RESOURCE_NAMES)
> +def test_rem_lockspace_parse_args(name):
> +    assert not util.sanlock_is_running(), "Test expects sanlock to be down"
>
> This will cause the test to fail, instead of error.
>
> Why not create a no_sanlock_daemon fixture, raising SanlockIsRunning
> exception
> if util.sanlock_is_running() returns True?
>
> @pytest.mark.parametrize("name", LOCKSPACE_OR_RESOURCE_NAMES)
> def test_rem_lockspace_parse_args(no_sanlock_daemon, name):
>     ...
>
> The current code duplicate the assert and the text for every test.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nirs/sanlock/pull/14?email_source=notifications&email_token=AL3BPNJXIRVQ7JOWRCII5B3PWQRXVA5CNFSM4HN7ALIKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOBZIIU6Y#pullrequestreview-240159355>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AL3BPNNMXPR5ULQNOH627KLPWQRXVANCNFSM4HN7ALIA>
> .
>

nirs commented on python/sanlock.c:
```diff
@@ -853,8 +853,8 @@ py_inq_lockspace(PyObject *self __unused, PyObject *args, PyObject *keywds)
-rem_lockspace(lockspace, host_id, path, offset=0, async=False, unused=False)\n\
-Remove a lockspace, releasing the acquired host_id. If async is True the\n\
+rem_lockspace(lockspace, host_id, path, offset=0, wait=True, unused=False)\n\
+Remove a lockspace, releasing the acquired host_id. If wait is False the\n\
```
In python 2 we should have:

    rem_lockspace(lockspace, host_id, path, offset=0, async=False, unused=False, wait=True)

In pyhton 3:

    rem_lockspace(lockspace, host_id, path, offset=0, unused=False, wait=True)

I think the best way to do this is to create a tiny wrapper for python 2 and python 3, which will do only the parameter parsing, and then call the common code doing the operation with normalized arguments.

We can register either the correct version in sanlock_methods list using ifdef.

nirs commented on python/sanlock.c:
```diff
@@ -864,30 +864,45 @@ successful termination these leases will be released.");
+        static char *legacy_kwlist[] = {"lockspace", "host_id", "path", "offset",
+                                            "async", "unused", NULL};
+        /* clear previous parse args errors */
+        PyErr_Clear();
```
We don't want to clear errors like this. We need 2 functions, each doing the right thing for the platform.

nirs commented on tests/python_test.py:
```diff
@@ -587,3 +587,13 @@ def test_reg_event_parse_args(no_sanlock_daemon, name):
+    ])
+def test_rem_lockspace_parse_args(no_sanlock_daemon, wait_key):
+    with raises_sanlock_errno():
+        sanlock.rem_lockspace(b"ls_name", 1, DUMMY_PATH, **{wait_key: False})
```
We need to test add_lockspace and remove lockspace with both async and wait on python 2. The best way would be to user parameters for the slow add_rem_lockspace and add_rem_lockspace_async, that will run both options on py2, and only wait= on py3.

aabawer replied:
that would slow tests x 2  for py2

nirs replied:
It will double the time for one or two tests, not all tests.

nirs reviewed (changes requested):
Please do not add new changes to this PR. Every PR should be focus on one topic.

This work should be deferred until we finish name and path handling.

aabawer commented:
Okay about the later PR.

Regarding py2:

why would we like to allow:

rem_lockspace(lockspace, host_id, path, offset=0, async=False,
unused=False, wait=True)

it makes sense to call:

rem_lockspace(lockspace, host_id, path, offset=0, unused=False, wait=True)
or
rem_lockspace(lockspace, host_id, path, offset=0, async=False, unused=False)

but not simultaneously even if wait=True and async=False and they do not
conflict.

nirs commented:
We don't wan to allow call with both arguments, but the docstring should that both exist.

If this was written in pure python, we would have something like:

```
def call_py2(async=None, wait=None, *args, **kw):
    """
    Arguments:
        async (bool): Return before operation was done. deprecated, use wait instead.
            Will be removed in python 3.
        wait (bool): Wait until operation is done.
    """
    if async is not None and wait is not None:
        raise ...

   if wait is None:
       wait = True if async is None else not async
   
    _real_call(wait, *args, **kw)

def call_py3(wait=True, *args, **kw):
   """
    Arguments:
        wait (bool): Wait until operation is done.
   """
    _real_call(wait, *args, **kw)
```

This example is probably wrong but I it explains the idea.

aabawer commented:
Thanks for clarifying.

On Wed, May 22, 2019 at 7:51 PM Nir Soffer <notifications@github.com> wrote:

> We don't wan to allow call with both arguments, but the docstring should
> that both exist.
>
> If this was written in pure python, we would have something like:
>
> def call_py2(async=None, wait=None, *args, **kw):
>     """
>     Arguments:
>         async (bool): Return before operation was done. deprecated, use wait instead.
>             Will be removed in python 3.
>         wait (bool): Wait until operation is done.
>     """
>     if async is not None and wait is not None:
>         raise ...
>
>    if wait is None:
>        wait = True if async is None else not async
>
>     _real_call(wait, *args, **kw)
>
> def call_py3(wait=True, *args, **kw):
>    """
>     Arguments:
>         wait (bool): Wait until operation is done.
>    """
>     _real_call(wait, *args, **kw)
>
> This example is probably wrong but I it explains the idea.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nirs/sanlock/pull/14?email_source=notifications&email_token=AL3BPNLFCS45SFJGO5TOML3PWV2Y3A5CNFSM4HN7ALIKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODV7VJLA#issuecomment-494884012>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AL3BPNKJWSTMZVH3SHUULJDPWV2Y3ANCNFSM4HN7ALIA>
> .
>

nirs commented on tests/conftest.py:
```diff
@@ -44,3 +44,12 @@ def user_4k_path(request):
+    if util.sanlock_is_running():
+        raise SanlockIsRunning
+
+class SanlockIsRunning(BaseException):
```
Exceptions should always inherit from Exception. Inheriting from BaseException is needed very rarely.

nirs commented on tests/conftest.py:
```diff
@@ -44,3 +44,12 @@ def user_4k_path(request):
+@pytest.fixture
+def no_sanlock_daemon():
+    if util.sanlock_is_running():
+        raise SanlockIsRunning
```
This works, but it would be nicer if the error expose the pid of the running process.

If we replace sanlock_is_running with:

    sanlock_status() -> {"pid": 1234, "running": True}

We can have a more helpful message here.

However I think this patch is good enough, so we can take it as is.

nirs commented on tests/util.py:
```diff
@@ -160,3 +160,22 @@ def generate_path(dirname, filename, encoding=None):
+    except OSError as e:
+        if e.errno != errno.ENOENT:
+            raise
+        return False
```
This return is not needed. If we do nothing, we will fall through to the next return.

nirs commented on tests/python_test.py:
```diff
@@ -448,6 +448,7 @@ def test_write_resource_invalid_align_sector(
     path = str(tmpdir.join("resources"))
     util.create_file(path, MIN_RES_SIZE)
     disks = [(path, 0)]
+    util.sanlock_is_running()
```
Leftover?

nirs commented on python/sanlock.c:
```diff
@@ -309,6 +309,47 @@ py_get_alignment(PyObject *self __unused, PyObject *args)
+ * as a Bytes object we raise TypeError with a message
+ * and return 0. Otherwise we convert arg into PyBytes
+ * and store the new object reference in addr while
+ * returning 1.
```
I thin it would be better to explain what we actually do:

- if the object is a PyUnicode object, create a new PyBytes object with the content of arg encoded as ascii.
if encoding fail will raise xxx.
- If the object is already a PyBytes object, increase the reference count and return it via addr pointer.
- Otherwise raise ValueError and return 0.

aabawer replied:
i think its slightly over-detailed for a relatively small piece of code. i'd stick with the general function description.

nirs replied:
The current version is misleading about converting PyBytes to PyBytes, lets make sure it is not.

nirs commented on python/sanlock.c:
```diff
@@ -309,6 +309,47 @@ py_get_alignment(PyObject *self __unused, PyObject *args)
+convert_to_pybytes(PyObject* arg, void *addr)
+{
+    if (arg == NULL)
+        return 1;
```
This is not possible, and we should crash if this happens.
We can do:

    assert(arg != NULL && "convert_to_pybytes called with NULL arg");

nirs commented on python/sanlock.c:
```diff
@@ -309,6 +309,47 @@ py_get_alignment(PyObject *self __unused, PyObject *args)
+
+    if (!PyBytes_Check(arg)) {
+        set_value_error("Argument type is not bytes or unicode: %s", arg);
+        return 0;
```
Why duplicate the code for checking PyBytes?

The check on line 342 works for both cases:

- py2, and arg is not unicode object
- py3

nirs commented on python/sanlock.c:
```diff
@@ -309,6 +309,47 @@ py_get_alignment(PyObject *self __unused, PyObject *args)
+    }
+
+    set_value_error("Argument type is not bytes: %s", arg);
+    return 0;
```
We promised to raise a TypeError. We can make it easy by converting set_value_error to:

    set_error(PyObject *err, const *char fmt, PyObject *obj)

nirs commented on python/sanlock.c:
```diff
@@ -347,10 +389,15 @@ py_init_lockspace(PyObject *self __unused, PyObject *args, PyObject *keywds)
+        goto exit_fail;
     }
 
+    Py_DECREF(lockspace);
```
Using exit_fail when we always need to cleanup duplicates the cleanup code, which lead to bugs when you forget to update both cleanup sections later.

I think it is better to use single return in these cases.

For example:

```
    int rv = -1;

    if (!parse_args(...))
        goto finally;

    rv = sanlock_direct_init(...);
    if (rv != 0)
        goto finally;

finally:
    Py_XDECREF(thing);
    if (rv != 0):
        return NULL;

    Py_RETURN_NONE;
```

nirs commented on python/sanlock.c:
```diff
@@ -309,6 +309,47 @@ py_get_alignment(PyObject *self __unused, PyObject *args)
+
 /* init_lockspace */
 PyDoc_STRVAR(pydoc_init_lockspace, "\
 init_lockspace(lockspace, path, offset=0, max_hosts=0, num_hosts=0, \
```
We can remove this in python 3 - is is deprecated for at least 5.5 years. Vdsm also stopped
using it few month ago, but we cannot break it in python 2.

Lets add a TODO for removing it in python 3, or file an issue in pagure.

nirs commented on python/sanlock.c:
```diff
@@ -398,11 +444,15 @@ py_init_resource(PyObject *self __unused, PyObject *args, PyObject *keywds)
     }
 
+    Py_DECREF(lockspace);
+    Py_DECREF(resource);
```
Double cleanup again ...

aabawer commented:
Thanks for reviewing. updated the PR.

On Thu, May 23, 2019 at 9:11 PM Nir Soffer <notifications@github.com> wrote:

> *@nirs* commented on this pull request.
> ------------------------------
>
> In python/sanlock.c
> <https://github.com/nirs/sanlock/pull/14#discussion_r287070353>:
>
> > @@ -309,6 +309,47 @@ py_get_alignment(PyObject *self __unused, PyObject *args)
>      return Py_BuildValue("i", rv);
>  }
>
> +/*
> + * converts parsed arg into PyBytes object.
> + * if arg is not covertable as a Python2 Unicode or
> + * as a Bytes object we raise TypeError with a message
> + * and return 0. Otherwise we convert arg into PyBytes
> + * and store the new object reference in addr while
> + * returning 1.
>
> The current version is misleading about converting PyBytes to PyBytes,
> lets make sure it is not.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nirs/sanlock/pull/14?email_source=notifications&email_token=AL3BPNI6O6CGGRGCA5H7GWTPW3M2ZA5CNFSM4HN7ALIKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOBZRPT4Y#discussion_r287070353>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AL3BPNPDVNAYYRATR3QRIL3PW3M2ZANCNFSM4HN7ALIA>
> .
>

nirs commented on python/sanlock.c:
```diff
@@ -347,10 +385,14 @@ py_init_lockspace(PyObject *self __unused, PyObject *args, PyObject *keywds)
+finally:
+    Py_XDECREF(lockspace);
+    if (rv == 0)
+        Py_RETURN_NONE;
```
I like to keep the happy path pattern - errors are check first, success is the normal path.

Keeping the same style everywhere helps to follow the code and avoid bugs.

nirs commented on python/sanlock.c:
```diff
@@ -395,14 +436,15 @@ py_init_resource(PyObject *self __unused, PyObject *args, PyObject *keywds)
+finally:
+    Py_XDECREF(lockspace);
+    Py_XDECREF(resource);
+    if (res) free(res);
```
free(NULL) is ok, we don't need the if, but looks like this is the way free() is used in this module, so we can improve this later.

nirs commented on python/sanlock.c:
```diff
@@ -737,9 +790,10 @@ acquired or released.");
+    int rv = -1, waitrs = 0, flags = 0;
+    const char *path;
     struct sanlk_lockspace ls;
+    PyObject *lockspace = NULL;
```
I think the order of the local variables should match the usage:

- input arguments (e.g. lockspace)
- internal variables (e.g. struct sanlk_lockspace ls)
- return value (e.g. rv)

The current code does not keep this convention, but at least we can add the new lockpace where it was defined before.

Keeping this order also requires having one declaration per line, which is a good thing, making it harder to create too many variables.

Since this patch is large lets just make sure the new lockspace/resource variables are located near the place they were before the patch.

nirs commented on python/sanlock.c:
```diff
@@ -943,6 +1004,7 @@ py_get_hosts(PyObject *self __unused, PyObject *args, PyObject *keywds)
     ls_list = __hosts_to_list(hss, hss_count);
 
 exit_fail:
```
For consistency, we should call this "finally" - this is not a failure path. Since this not an issue added by the patch we can fix it later.

nirs commented on python/sanlock.c:
```diff
@@ -1527,29 +1603,33 @@ with -EBUSY error in this case.\n\
+finally:
+    Py_XDECREF(lockspace);
+    if (rv < 0)
+        return NULL;
```
This is the pattern we need everywhere.

nirs commented on tests/python_test.py:
```diff
@@ -54,16 +54,15 @@
     pytest.param(
         u"ascii",
-        marks=pytest.mark.skipif(six.PY3, reason="python 3 supports only bytes")),
+        marks=pytest.mark.xfail(six.PY3, reason="python 3 only supports bytes")),
```
This is not expected failure, this input value is not valid on python 3, this is why we skip it. Lets keep the skip.

If you want to test behaviour with unicode names, you can add tests for raising TypeError, but I don't think it worth the effort. Maybe just one test to make sure convert_to_pybytes behaves correctly?

aabawer replied:
IMO we expect py3 to have TypeError here for unicode input in current implementation. If it would XPASS for some reason it would hint us there is something wrong and undocumented with lockspace/resource parsing. If we skip it by default, we'll probably never find it.

nirs replied:
xfail is not mean for testing negative flows. It should be used to mark a test as expected failure until we have time to fix the code. Here there is nothing to fix, so we skip the test.

If you want to test that we raise TypeError, add one test calling with invalid type, and check that we raise the expected error.

This parameter will be removed once we drop python 2, but your negative test about type error should always run.

nirs reviewed (changes requested):
Thanks, merged all patches except the last 2 to my py3 branch. Please rebase.

aabawer commented:
Thanks, pushed fixes.

nirs commented on python/sanlock.c:
```diff
@@ -149,7 +149,7 @@ __parse_resource(PyObject *obj, struct sanlk_resource **res_ret)
     *res_ret = res;
     return 0;
```
This function does not use try-finally pattern, but try-except. If we fail, we go to exit_fail and free res. If we succeed, we return res.

nirs commented on python/sanlock.c:
```diff
@@ -235,18 +235,18 @@ __hosts_to_list(struct sanlk_host *hss, int hss_count)
         Py_DECREF(ls_entry);
     }
 
     return ls_list;
```
Same, we need to decref ls_list and ls_entry only on error, so this is not a finaly.

nirs commented on python/sanlock.c:
```diff
@@ -592,17 +639,12 @@ py_read_resource(PyObject *self __unused, PyObject *args, PyObject *keywds)
-    /* success */
-    free(rs);
+finally:
+    if(rs) free(rs);
```
if() -> if ()

nirs commented on python/sanlock.c:
```diff
@@ -592,17 +639,12 @@ py_read_resource(PyObject *self __unused, PyObject *args, PyObject *keywds)
         "version", rs->lver);
     if (rs_info  == NULL)
-        goto exit_fail;
+        goto finally;
```
We can remove this check and goto now.

nirs commented on python/sanlock.c:
```diff
@@ -894,7 +951,7 @@ py_get_lockspaces(PyObject *self __unused, PyObject *args, PyObject *keywds)
     return ls_list;
```
This is not a finally pattern.

nirs commented on python/sanlock.c:
```diff
@@ -1316,22 +1378,28 @@ unregister the event listener using end_event.");
+        goto finally;
     }
 
     return Py_BuildValue("i", fd);
```
This is not finally pattern.

nirs commented on python/sanlock.c:
```diff
@@ -1316,22 +1378,28 @@ unregister the event listener using end_event.");
     return Py_BuildValue("i", fd);
+
+finally:
+    Py_XDECREF(lockspace);
```
Because this function does not go to finally in all cases, we leak the lockspace here.

nirs commented on python/sanlock.c:
```diff
@@ -1376,70 +1444,70 @@ py_get_event(PyObject *self __unused, PyObject *args)
     return events;
 
-exit_fail:
+finally:
```
This is not finally pattern

nirs reviewed (changes requested):
This patch is way to big and consume too much time to review.

I think this should be separated to these changes:
- change exit_fail used as finally to finally (but keep exit_fail used only on failures)
- add convert_to_pybytes function
- tests for new behavior, possibly marked as xfail when needed
- the implementation of the easy functions
- the more tricky functions, possibly one patch for each tricky function.

The trivial test refactoring adding constant is not related to this PR, please send it later when we finish with this work.

aabawer commented:
I agree that we dont have to make such a big patch.
In fact i think that we dont have to change the exit_fail/finally in most
occurrences as we usually have the following flow:
1. parse lockspace/resource/path
2. check for parse failure - here we should xdecref for safety
3. set parameters for sanlock api call - here we can decref any parsed
object after setup as we are done with those parsed objects
4. call sanock api

On Fri, May 31, 2019 at 2:12 AM Nir Soffer <notifications@github.com> wrote:

> *@nirs* requested changes on this pull request.
>
> This patch is way to big and consume too much time to review.
>
> I think this should be separated to these changes:
>
>    - change exit_fail used as finally to finally (but keep exit_fail used
>    only on failures)
>    - add convert_to_pybytes function
>    - tests for new behavior, possibly marked as xfail when needed
>    - the implementation of the easy functions
>    - the more tricky functions, possibly one patch for each tricky
>    function.
>
> The trivial test refactoring adding constant is not related to this PR,
> please send it later when we finish with this work.
> ------------------------------
>
> In python/sanlock.c
> <https://github.com/nirs/sanlock/pull/14#discussion_r289193039>:
>
> > @@ -149,7 +149,7 @@ __parse_resource(PyObject *obj, struct sanlk_resource **res_ret)
>      *res_ret = res;
>      return 0;
>
> This function does not use try-finally pattern, but try-except. If we
> fail, we go to exit_fail and free res. If we succeed, we return res.
> ------------------------------
>
> In python/sanlock.c
> <https://github.com/nirs/sanlock/pull/14#discussion_r289193484>:
>
> >
>          if (PyList_Append(ls_list, ls_entry) != 0)
> -            goto exit_fail;
> +            goto finally;
>
>          Py_DECREF(ls_entry);
>      }
>
>      return ls_list;
>
> Same, we need to decref ls_list and ls_entry only on error, so this is not
> a finaly.
> ------------------------------
>
> In python/sanlock.c
> <https://github.com/nirs/sanlock/pull/14#discussion_r289197870>:
>
> >
> -    /* success */
> -    free(rs);
> +finally:
> +    if(rs) free(rs);
>
> if() -> if ()
> ------------------------------
>
> In python/sanlock.c
> <https://github.com/nirs/sanlock/pull/14#discussion_r289198387>:
>
> > @@ -592,17 +639,12 @@ py_read_resource(PyObject *self __unused, PyObject *args, PyObject *keywds)
>          "resource", rs->name,
>          "version", rs->lver);
>      if (rs_info  == NULL)
> -        goto exit_fail;
> +        goto finally;
>
> We can remove this check and goto now.
> ------------------------------
>
> In python/sanlock.c
> <https://github.com/nirs/sanlock/pull/14#discussion_r289199893>:
>
> > @@ -894,7 +951,7 @@ py_get_lockspaces(PyObject *self __unused, PyObject *args, PyObject *keywds)
>      return ls_list;
>
> This is not a finally pattern.
> ------------------------------
>
> In python/sanlock.c
> <https://github.com/nirs/sanlock/pull/14#discussion_r289202218>:
>
> >      Py_END_ALLOW_THREADS
>
>      if (fd < 0) {
>          __set_exception(fd, "Unable to register event fd");
> -        return NULL;
> +        goto finally;
>      }
>
>      return Py_BuildValue("i", fd);
>
> This is not finally pattern.
> ------------------------------
>
> In python/sanlock.c
> <https://github.com/nirs/sanlock/pull/14#discussion_r289202497>:
>
> >      }
>
>      return Py_BuildValue("i", fd);
> +
> +finally:
> +    Py_XDECREF(lockspace);
>
> Because this function does not go to finally in all cases, we leak the
> lockspace here.
> ------------------------------
>
> In python/sanlock.c
> <https://github.com/nirs/sanlock/pull/14#discussion_r289202929>:
>
> >
>          Py_DECREF(item);
>          item = NULL;
>      }
>
>      return events;
>
> -exit_fail:
> +finally:
>
> This is not finally pattern
> ------------------------------
>
> In tests/python_test.py
> <https://github.com/nirs/sanlock/pull/14#discussion_r289204511>:
>
> >      # Python 2 also supports str.
>      pytest.param(
>          "\xd7\x90",
>          marks=pytest.mark.skipif(six.PY3, reason="python 3 supports only bytes")),
>      # Python 2 also supports unicode with ascii content.
>      pytest.param(
>          u"ascii",
> -        marks=pytest.mark.skipif(six.PY3, reason="python 3 supports only bytes")),
> +        marks=pytest.mark.xfail(six.PY3, reason="python 3 only supports bytes")),
>
> xfail is not mean for testing negative flows. It should be used to mark a
> test as expected failure until we have time to fix the code. Here there is
> nothing to fix, so we skip the test.
>
> If you want to test that we raise TypeError, add one test calling with
> invalid type, and check that we raise the expected error.
>
> This parameter will be removed once we drop python 2, but your negative
> test about type error should always run.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nirs/sanlock/pull/14?email_source=notifications&email_token=AL3BPNKW7KH4WWIB4WVFPVTPYBNPPA5CNFSM4HN7ALIKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB2GAAHA#pullrequestreview-244056092>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AL3BPNJNRCZ4RIC27RLAXQ3PYBNPPANCNFSM4HN7ALIA>
> .
>

aabawer commented:
btw, we can use "__attribute__((cleanup))" for Internal PyObjects . its allowed since we already have __unused in code for "self" in api declarations, so the entire python/sanlock.c code is in pure gcc.
i think its a better option than using the "goto exit_fail" method.

nirs commented:
The issue is no change from exit_fail to finally, or using or not using gcc attributes, but trying to make the entire change at once in messy code.

For example having exit_fail that behave like finally, having multiple variables declarations per line, having double cleanup code. Fixing these issues before we add new code will make it much easier and safer to change the code

Getting rid of the temporary bytes object at the start of the function is require having cleanup twice, once for parsing, and then again after copying the values to sanlock structures. I don't want to have double cleanup code, this makes it harder to change the code.

Using more gcc attributes adds more complexity and risk, and may break with other compilers. We should use the same solution used by the upstream python source.

The solution suggested in the current code is good, we just need fix the incorrect changes, and split it to easy to review chunks. For example once the tests are correct, I can merge them and we don't have to 
continue reviewing them. But it the tests are part of the same huge commit I need to review them again each time.

aabawer commented:
okay

On Sun, Jun 2, 2019 at 11:33 AM Nir Soffer <notifications@github.com> wrote:

> The issue is no change from exit_fail to finally, or using or not using
> gcc attributes, but trying to make the entire change at once in messy code.
>
> For example having exit_fail that behave like finally, having multiple
> variables declarations per line, having double cleanup code. Fixing these
> issues before we add new code will make it much easier and safer to change
> the code
>
> Getting rid of the temporary bytes object at the start of the function is
> require having cleanup twice, once for parsing, and then again after
> copying the values to sanlock structures. I don't want to have double
> cleanup code, this makes it harder to change the code.
>
> Using more gcc attributes adds more complexity and risk, and may break
> with other compilers. We should use the same solution used by the upstream
> python source.
>
> The solution suggested in the current code is good, we just need fix the
> incorrect changes, and split it to easy to review chunks. For example once
> the tests are correct, I can merge them and we don't have to
> continue reviewing them. But it the tests are part of the same huge commit
> I need to review them again each time.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nirs/sanlock/pull/14?email_source=notifications&email_token=AL3BPNLQNYIPNQTQ4XZR5YTPYOAUXA5CNFSM4HN7ALIKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWXQV2Q#issuecomment-498010858>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AL3BPNPM7UCJEB3ULLYXC4LPYOAUXANCNFSM4HN7ALIA>
> .
>

aabawer commented:
will deliver a fresh PR
